### DISCUSSION TITLE
Recommended patterns for React 19 / React Router v7

### DISCUSSION THREAD

**Discussion by pspeter3** [remix-run/react-router#12492] [Q&A]
Now that React 19 also supports Form Actions, when do you use `action` / `clientAction` from React Router vs a React form `action`?

**Comment by sergiodxa** [comment]
The actions generate a URL and also run server side, until RR supports React server action the form actions are just a better onSubmit

**Reply by pspeter3** [reply]
That's fair. I didn't know if the `useOptimistic` or other new hooks might be helpful
